File: src/Java/array/85.最大矩形.java
/*
 * @lc app=leetcode.cn id=85 lang=java
 *
 * [85] 最大矩形
 */

// @lc code=start
class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix.length == 0) return 0;
        int[] heights = new int[matrix[0].length + 1];
        int res = 0;
        for (int i = 0; i < matrix.length; i++) {
            Deque<Integer> stack = new ArrayDeque<>();
            heights[matrix[0].length] = 0;
            // ! until maxtrix[0].length
            for (int j = 0; j <= matrix[0].length; j++) {
                // count the height of the bar => transfer to problem 84
                if (j < matrix[0].length) {
                    if (matrix[i][j] == '1') heights[j]++;
                    else heights[j] = 0;
                }
                // problem 84(count the max matrix row by row)
                while (!stack.isEmpty() && heights[j] < heights[stack.peek()]) {
                    int cur = stack.pop();
                    int leftLessMin = stack.isEmpty()? -1 : stack.peek();
                    res = Math.max(res, (j - leftLessMin - 1) * heights[cur]);
                }
                stack.push(j);
            }
        }
        return res;
    }
}
// @lc code=end

File: src/Python/array/85.最大矩形.py
#
# @lc app=leetcode.cn id=85 lang=python
#
# [85] 最大矩形
#

# @lc code=start
class Solution(object):
    def maximalRectangle(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        if len(matrix) == 0: return 0
        # to make the last element of heights array can be calculated
        heights = [0] * (len(matrix[0]) + 1)
        res = 0
        for i in range(len(matrix)):
            stack = []
            heights[len(matrix[0])] = 0
            # get the heights array
            for j in range(len(matrix[0]) + 1):
                if j < len(matrix[0]):
                    if matrix[i][j] == '1':
                        heights[j] += 1
                    else:
                        heights[j] = 0
                # problem 85 (transfer to count max rectangular row by row)
                while stack and heights[j] < heights[stack[-1]]:
                    cur = stack.pop()
                    leftMin = stack[-1] if stack else -1
                    res = max(res, (j - leftMin - 1) * heights[cur])
                stack.append(j)
        return res
# @lc code=end

